test/scripts/run_cuda_memcheck.py

```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env python
+
+"""This script runs cuda-memcheck on the specified unit test. Each test case
+is run in its isolated process with a timeout so that:
+1) different test cases won't influence each other, and
+2) in case of hang, the script would still finish in a finite amount of time.
+The output will be written to a log file result.log
+
+Example usage:
+    python run_cuda_memcheck.py ../test_torch.py 600
+
+Note that running cuda-memcheck could be very slow.
+"""
+
+import asyncio
+import torch
+import multiprocessing
+import argparse
+import subprocess
+import tqdm
+import re
+
+ALL_TESTS = []
+NUM_PROCESSES = multiprocessing.cpu_count()
+GPUS = torch.cuda.device_count()
+
+# parse arguments
+parser = argparse.ArgumentParser(description="Run isolated cuda-memcheck on unit tests")
+parser.add_argument('filename', help="the python file for a test, such as test_torch.py")
+parser.add_argument('timeout', type=int, help='kill the test if it does not terminate in a certain amount of seconds')
+parser.add_argument('--ignore', nargs='+', default=['cudaErrorInvalidDeviceFunction'],
+                    help='list of regex of the failures not interested, default to ["cudaErrorInvalidDeviceFunction"], '
+                         'because cublas does not run error-free under cuda-memcheck')
+args = parser.parse_args()
+
+# Discover tests:
+# To get a list of tests, run:
+# pytest --setup-only test/test_torch.py
+# and then parse the output
+proc = subprocess.Popen(['pytest', '--setup-only', args.filename], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+stdout, stderr = proc.communicate()
+lines = stdout.decode().strip().splitlines()
+for line in lines:
+    if '(fixtures used:' in line:
+        line = line.strip().split()[0]
+        line = line[line.find('::') + 2:]
+        line = line.replace('::', '.')
+        ALL_TESTS.append(line)
+
+# Run tests:
+# Since running cuda-memcheck on PyTorch unit tests is very slow, these tests must be run in parallel.
+# This is done by using the coroutine feature in new Python versions.  A number of coroutines are created;
+# they create subprocesses and awaiting them to finish. The number of running subprocesses is the same as
+# the number of CPUs in the machine. These subprocesses are balanced across different GPUs on the system.
+progress = 0
+logfile = open('result.log', 'w')
+progressbar = tqdm.tqdm(total=len(ALL_TESTS))
+
+async def run1():
+    global progress
+    while progress < len(ALL_TESTS):
+        test = ALL_TESTS[progress]
+        progress += 1
+        cmd = f'CUDA_VISIBLE_DEVICES={progress % GPUS} cuda-memcheck --error-exitcode 1 python {args.filename} {test}'
```

Comment:
Should you be assigning 1 GPU per async process instead of rotating the GPUs using progress? That could mean that some jobs are running concurrently on two GPUs, right?

Should we make the number of CUDA devices per async call configurable so we can run multi-gpu tests, too?

Reply:
Currently, no single job can run on multiple GPUs, but multiple jobs could run on the same GPU. I have made it configurable so that users could run on multigpu jobs.